src/test/java/com/crowdin/cli/commands/actions/FileUploadActionTest.java

```diff
@@ -0,0 +1,145 @@
+
+        NewAction<ProjectProperties, ProjectClient> action = new FileUploadAction(fileToUpload, null, false, null, null, false, false, null, false);
+        action.act(Outputter.getDefault(), pb, client);
+
+        verify(client).downloadFullProject(any());
+        verify(client).uploadStorage(eq("first.po"), any());
+        AddFileRequest addFileRequest = new AddFileRequest() {{
+            setName("first.po");
+            setStorageId(1L);
+        }};
+        verify(client).addSource(eq(addFileRequest));
+        verifyNoMoreInteractions(client);
+    }
+
+    @Test
+    public void testUpload_StringBasedProject() throws ResponseException {
+        File fileToUpload = new File(project.getBasePath() + "first.po");
+        project.addFile(Utils.normalizePath("first.po"), "Hello, World!");
+        NewPropertiesWithFilesUtilBuilder pbBuilder = NewPropertiesWithFilesUtilBuilder
+            .minimalBuiltPropertiesBean("*", Utils.PATH_SEPARATOR + "%original_file_name%-CR-%locale%")
+            .setBasePath(project.getBasePath());
+        PropertiesWithFiles pb = pbBuilder.build();
+        ProjectClient client = mock(ProjectClient.class);
+        Branch branch = mock(Branch.class);
+        CrowdinProjectFull build = ProjectBuilder.emptyProject(Long.parseLong(pb.getProjectId())).build();
+        build.setType(Type.STRINGS_BASED);
+        UploadStringsProgress progress = new UploadStringsProgress() {{
+            setIdentifier("id");
+            setProgress(100);
+        }};
+        UploadStringsProgress progressFinished = new UploadStringsProgress() {{
+            setIdentifier("id");
+            setStatus("finished");
+            setProgress(100);
+        }};
+
+        when(branch.getId()).thenReturn(2L);
+        when(client.downloadFullProject(any())).thenReturn(build);
+        when(client.uploadStorage(eq("first.po"), any())).thenReturn(1L);
+        when(client.addBranch(any())).thenReturn(branch);
+        when(client.addSourceStringsBased(any())).thenReturn(progress);
+        when(client.getUploadStringsStatus(any())).thenReturn(progressFinished);
+
+        NewAction<ProjectProperties, ProjectClient> action = new FileUploadAction(fileToUpload, "branch", false, null, null, false, false, null, false);
+        action.act(Outputter.getDefault(), pb, client);
+
+        verify(client).downloadFullProject(any());
+        verify(client).uploadStorage(eq("first.po"), any());
+        verify(client).addBranch(any());
+        verify(client).getUploadStringsStatus(any());
+        UploadStringsRequest addFileRequest = new UploadStringsRequest() {{
+            setBranchId(2L);
+            setStorageId(1L);
+            setCleanupMode(false);
+            setUpdateStrings(false);
+        }};
+        verify(client).addSourceStringsBased(eq(addFileRequest));
+        verifyNoMoreInteractions(client);
+    }
+
+    @Test
+    public void testUploadUpdate_FileBasedProject() throws ResponseException {

```

Comment:
\+ separate test for `--no-auto-update`

Reply:
\+ a test for uploading a file to a branch when the branch is missing in Crowdin

Reply:
\+ a separate case to test the `--labels` and `--excluded-language` options

Reply:
\+ a test for the `--dest` parameter (including directories, to check that command creates directories and uploads a file to the specified place)